✨ Add support for copying `contextvars` to thread workers

tiangolo opened the pull request:
✨ Add support for copying `contextvars` to thread workers.

This is my attempt at implementing what's described here: https://github.com/python-trio/trio/issues/648

The original post said:

> * `run_sync_in_worker_thread` should do `copy_context` in the trio thread and then `ctx.run(...)` in the worker thread
>   * And it should unset the `sniffio` magic contextvar in the child
> * `BlockingTrioPortal` should do `copy_context` in the blocking thread, and then use the context in Trio
>   * But Trio doesn't currently have a way to say "use this context for this task", so I guess we'd need to add that. A `context=` argument to `spawn_system_task` would work, and we might as well add it to `start_soon` while we're at it.
>     * I guess these will also want to set up the `sniffio` magic contextvar. Maybe they should always copy the context that's passed in, and then mutate the copy?

I looked at the release notes and realized these names have been deprecated/replaced by other things:

`run_sync_in_worker_thread` -> `to_thread.run_sync`
`BlockingTrioPortal` -> `from_thread.run_sync` and `from_thread.run`

I'm also assuming that the "`sniffio` magic contextvar" refers to `sniffio.current_async_library_cvar`.

richardsheridan commented on docs/source/reference-core.rst:
```diff
@@ -1021,8 +1021,18 @@ Example output (yours may differ slightly):
+    In advanced use cases you can also pass a custom `contextvars` context to
+    ``nursery.start_soon(child, context=context)``.
+
+    That's an advanced use case, and you probably won't need that easily, but now you
+    know in case you need that in the future.
```
```suggestion
    That's an advanced use case that you probably won't need, but now you
    know why that keyword argument exists!
```

richardsheridan commented on docs/source/reference-core.rst:
```diff
@@ -1820,6 +1830,66 @@ to spawn a child thread, and then use a :ref:`memory channel
+will be available for that function and other internal/children tasks, but the value
+won't be available in the parent thread.
+
+If you need to modify values that would live in the context variables and you need to
+make those modifications from the children threads, you can instead set an object
+(e.g. a dictionary) in the context variable, in the top level/parent Trio thread. And
+then in the children, instead of setting the context variable, you can ``get`` the same
+object, and modify its values. That way you keep the same object in the context
+variable and only mutate it in children threads.
+
+This way, you can modify the object content in children threads and still access the
+new content in the parent thread.
```
```suggestion
If you need to modify values that would live in the context variables and you need to
make those modifications from the child threads, you can instead set a mutable object
(e.g. a dictionary) in the context variable of the top level/parent Trio thread.
Then in the children, instead of setting the context variable, you can ``get`` the same
object, and modify its values. That way you keep the same object in the context
variable and only mutate it in child threads.

This way, you can modify the object content in child threads and still access the
new content in the parent thread.
```

richardsheridan reviewed (approved):
Just some non-critical wording nitpicks; I like the overall state of the PR. Still, I would like to wait for @njsmith to weigh in on the issue before changing the call signature of `Nursery.start_soon`!

tiangolo commented:
Awesome, thanks for the review @richardsheridan! I agree, I would love to have @njsmith's point of view on this!

richardsheridan commented:
If it were just me, I would merge the PR for the next release but without the kwarg change to `Nursery.start_soon` because it's redundant to `context.run(nursery.start_soon)` and a function signature change in the `lowlevel` package only is much more tolerable than in the core API.  Also for what it's worth, it would keep things in sync with @agronholm choices in AnyIO.

agronholm commented on newsfragments/2160.feature.rst:
```diff
@@ -0,0 +1,6 @@
+Now context variables set with `contextvars` are preserved when runing functions
```
```suggestion
Now context variables set with `contextvars` are preserved when running functions
```

agronholm commented:
> If it were just me, I would merge the PR for the next release but without the kwarg change to `Nursery.start_soon` because it's redundant to `context.run(nursery.start_soon)` and a function signature change in the `lowlevel` package only is much more tolerable than in the core API. Also for what it's worth, it would keep things in sync with @agronholm choices in AnyIO.

I fully agree. @tiangolo would it be possible to modify this PR or create another one without the API changes?

agronholm commented on newsfragments/2160.feature.rst:
```diff
@@ -0,0 +1,6 @@
+in a worker thread with `trio.to_thread.run_sync`, or when running
+functions from the worker thread in the parent Trio thread with
+`trio.from_thread.run`, and `trio.from_thread.run_sync`.
+This is done by automatically copying the `contextvars` context. You can also pass a
+custom `contextvars` context to ``nursery.start_soon(child, context=context)``.
```
This section needs to be updated.
```suggestion
This is done by automatically copying the `contextvars` context``.
```

tiangolo commented:
Awesome, thanks for the feedback @richardsheridan and @agronholm! :bowing_man: :rocket: 

Done, I updated it accordingly. Let me know if you see anything else or if I missed something.

richardsheridan commented:
The failing tests will likely be fixed by merging master.

agronholm commented on trio/_core/_run.py:
```diff
@@ -1540,7 +1543,7 @@ def task_exited(self, task, outcome):
 
     @_public
-    def spawn_system_task(self, async_fn, *args, name=None):
+    def spawn_system_task(self, async_fn, *args, name=None, context=None):
```
I don't think this is supposed to be here anymore? System tasks should always inherit the system context, right?

tiangolo replied:
I actually haven't been able to make tests pass without it, I've been trying several different approaches, copying the context in different places, but in some I get errors that show it's not copying the right context, or thread timeout errors, or it breaks internals. :scream: I suspect it could be just that I lack the knowledge of the internals to do it in a different and better way.

Do you see a way that could be solved to remove it?

On the other hand, I understand @richardsheridan thinks this could be useful... for me it's the same, but I just don't see how to make it work without this. :thinking:

richardsheridan commented:
I think that's a good feature, otherwise there is no way to inject any
other context to a system task...

On Sun, Jan 30, 2022, 11:10 AM Alex Grönholm ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In trio/_core/_run.py
> <https://github.com/python-trio/trio/pull/2160#discussion_r795208447>:
>
> > @@ -1540,7 +1543,7 @@ def task_exited(self, task, outcome):
>      ################
>
>      @_public
> -    def spawn_system_task(self, async_fn, *args, name=None):
> +    def spawn_system_task(self, async_fn, *args, name=None, context=None):
>
> I don't think this is supposed to be here anymore? System tasks should
> always inherit the system context, right?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/python-trio/trio/pull/2160#pullrequestreview-867140324>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AB5BHRSUGNPVHRB3ACB37SLUYVPGLANCNFSM5G3IHZUQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

tiangolo commented:
I just noticed this warning in `_core/_generated_run.py`: https://github.com/python-trio/trio/blob/master/trio/_core/_generated_run.py#L1-L3

But I can't see what generates it, the only reference to `generated_run` I get searching in my editor is the import in `_run.py`, do you know what's that note about? Or what should I modify instead?

richardsheridan commented:
Right you should run [`gen_exports.py`](https://github.com/python-trio/trio/blob/master/trio/_tools/gen_exports.py) after making changes to public apis in `_run.py` and some others. It seemed like you were already doing that!

agronholm commented:
> I think that's a good feature, otherwise there is no way to inject any other context to a system task...
> […](#)
> On Sun, Jan 30, 2022, 11:10 AM Alex Grönholm ***@***.***> wrote: ***@***.**** commented on this pull request. ------------------------------ In trio/_core/_run.py <[#2160 (comment)](https://github.com/python-trio/trio/pull/2160#discussion_r795208447)>: > @@ -1540,7 +1543,7 @@ def task_exited(self, task, outcome): ################ @_public - def spawn_system_task(self, async_fn, *args, name=None): + def spawn_system_task(self, async_fn, *args, name=None, context=None): I don't think this is supposed to be here anymore? System tasks should always inherit the system context, right? — Reply to this email directly, view it on GitHub <[#2160 (review)](https://github.com/python-trio/trio/pull/2160#pullrequestreview-867140324)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AB5BHRSUGNPVHRB3ACB37SLUYVPGLANCNFSM5G3IHZUQ> . You are receiving this because you were mentioned.Message ID: ***@***.***>

That API addition is not mentioned in the news fragment.

tiangolo commented:
> That API addition is not mentioned in the news fragment.

@agronholm get it, just updated it. I suspect only about `trio.lowlevel.spawn_system_task`, right? That's what I added to the newsfragment.

richardsheridan commented on newsfragments/2160.feature.rst:
```diff
@@ -0,0 +1,6 @@
+functions from the worker thread in the parent Trio thread with
+`trio.from_thread.run`, and `trio.from_thread.run_sync`.
+This is done by automatically copying the `contextvars` context.
+`trio.lowlevel.spawn_system_task` now also receives an optional `context` argument.
```
Context needs double backticks

```suggestion
`trio.lowlevel.spawn_system_task` now also receives an optional ``context`` argument.
```

tiangolo replied:
Ha! I just noticed that and was about to commit it, thanks! 😅

tiangolo commented:
Thanks @pquentin! :rocket:

oremanj commented:
This PR introduced a regression: calling `spawn_system_task` now changes the current_async_library_cvar in the _calling_ task too, even if that's in a different async library such as asyncio. This violates the rule that it is generally fine to do synchronous cross-calls between two event loops in the same thread without additional ceremony.

richardsheridan commented:
> This PR introduced a regression: calling `spawn_system_task` now changes the current_async_library_cvar in the _calling_ task too, even if that's in a different async library such as asyncio. This violates the rule that it is generally fine to do synchronous cross-calls between two event loops in the same thread without additional ceremony.

For my future self's reference: This was regression resulted in #2183 and was fixed in #2462.
